File: GOF/Bridge/java/src/Window/Window.java
package Window;

import Elements.Point;
import Elements.View;
import Factory.WindowSystemFactory;
import Window.WindowImp.WindowImp;

public class Window {

	private WindowImp _imp;
	private View _content;

	public Window(View content) {
		_content = content;
	}

	public void drawContents() {}
	public void open() {}
	public void iconify() {}
	public void deiconify() {}
	public void setOrigin(Point at) {}
	public void setExtent(Point extent) {}
	public void lower() {}
	public void drawLine(Point s, Point e) {}
	public void drawRect(Point bottomLeft, Point topRight) {}
	public void drawPolgen(Point[] points) {}
	public void drawText(String text, Point position) {}

	protected WindowImp getWindowImp() {
		if(_imp == null)
			_imp = WindowSystemFactory.getInstance().makeWindowImp();
		return _imp;
	}

	protected View getView() {
		return _content;
	}
}

File: GOF/Bridge/cpp/Window/Window.cpp
//
// Created by 17271 on 2019/3/29.
//

#include "Window.h"
#include "WindowImp/WindowImp.h"
#include "../WindowSystemFactory.h"

using std::weak_ptr;

Window::Window(std::weak_ptr<View> contents)
	: _contents{contents}
{}

void Window::drawContents()
{}

void Window::open()
{}

void Window::close()
{}

void Window::iconify()
{}

void Window::deiconify()
{}

void Window::setOrigin(Point &at)
{}

void Window::setExtent(Point &extent)
{}

void Window::lower()
{}

void Window::drawLine(Point &s, Point &e)
{}

void Window::drawRect(Point &bottomLeft, Point &topRight)
{
	weak_ptr<WindowImp> imp = getWindowImp();
	imp.lock()->deviceRect(bottomLeft.x(), bottomLeft.y(), topRight.x(), topRight.y());
}

void Window::drawPolgen(std::vector<Point> points)
{}

void Window::drawText(std::string text, Point &position)
{}

std::weak_ptr<WindowImp> Window::getWindowImp()
{
	if(_imp == nullptr)
		_imp = WindowSystemFactory::getInstance().makeWindowImp();
	return _imp;
}

std::weak_ptr<View> Window::getView()
{
	return _contents;
}